include/cudaq/Optimizer/Dialect/Quake/Canonical.h

```diff
@@ -0,0 +1,125 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#pragma once
+
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeTypes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/PatternMatch.h"
+
+namespace quake::canonical {
+
+inline mlir::Value createCast(mlir::PatternRewriter &rewriter,
+                              mlir::Location loc, mlir::Value inVal) {
+  auto i64Ty = rewriter.getI64Type();
+  assert(inVal.getType() != rewriter.getIndexType() &&
+         "use of index type is deprecated");
```

Comment:
It looks like this code was moved, and it was "deprecated" before. Should we promote it to full-on "unsupported" now?
```suggestion
         "use of index type is unsupported");
```